### DISCUSSION TITLE
Is it expected that Vitest should not require/resolve `.../index.ts` paths?

### DISCUSSION THREAD

**Discussion by moltar** [vitest-dev/vitest#5201] [Q&A] [answered]
E.g.:

We have a file `./foo/index.ts`

Trying to solve it this way:

```ts
require.resolve('./foo')
```

Fails with:

> Cannot find module './foo'

---

But it works fine this way:

```ts
require.resolve('./foo/index.ts')
```

---

From what I read, Vitest claims to go by Node.js resolutions, which I am guessing then implies no `index.ts` as the directory index, given that Node.js is not TS-aware.

Is that true?

If so, are there any workarounds or configurations I could change to make Vitest understand `index.ts` indexes?

**Comment by hi-ogawa** [comment, marked as answer]
> From what I read, Vitest claims to go by Node.js resolutions, which I am guessing then implies no index.ts as the directory index, given that Node.js is not TS-aware.
> 
> Is that true?

Yes, it's expected. Vitest doesn't intercept what happens within `require`, so you'll likely get the same behavior as NodeJs.

EDIT: Ah, maybe not. Vitest might be mutating `require.extensions` in some cases

https://github.com/vitest-dev/vitest/blob/cad955b117fbd81ef623fd8829b98f93b00dae26/packages/vitest/src/runtime/setup-node.ts#L36-L41

So, you might try something similar `require.extensions[".ts"] = () => {}` in `setupFiles` or somewhere but I don't know about the consequences.

> If so, are there any workarounds or configurations I could change to make Vitest understand index.ts indexes?

To help finding a workaround, we would need a little more context of the overall usage. Otherwise, you've already found a workaround to use `require.resolve('./foo/index.ts')`.

I assume you can use `require('./foo')` outside Vitest since you use custom loader or you transpile the file into `foo/index.js` during runtime.

**Reply by moltar** [reply]
Thank you for the reply and confirmation @hi-ogawa!

> I assume you can use require('./foo') outside Vitest since you use custom loader or you transpile the file into foo/index.js during runtime.

We use [`tsx`](https://github.com/privatenumber/tsx) to execute the TypeScript code and it works fine. As I understand `tsx` is also `esbuild`-based, but, perhaps, it has adds to the resolution process to include `.ts` extensions?

**Reply by moltar** [reply]
Ah, good lean on the `require.extensions` link!

Found the issue discussing this: https://github.com/vitest-dev/vitest/issues/846
